### PR TITLE
fix: Use mechanism meta for error grouping

### PR DESCRIPTION
**Pull request by philipphofmann**
## :scroll: Description

Instead of using the SDK fingerprint use the mechanism for error grouping.
The value of the exception now contains both value and localizedDescription.
This change affects grouping if the server is not patched with the special
handling of grouping for NSErrors.

Only release after https://github.com/getsentry/sentry/pull/23794 is deployed.

## :bulb: Motivation and Context

https://github.com/getsentry/sentry-cocoa/discussions/929#discussioncomment-359377. Is related to https://github.com/getsentry/sentry/pull/23794.

## :green_heart: How did you test it?
Unit tests and Simulator.

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [x] I updated the CHANGELOG if needed
- [ ] I updated the docs if needed
- [x] Review from the native team if needed
- [x] No breaking changes

### CODE REVIEW

**Comment by bruno-garcia** [comment]
Tried to fix the conflict but I probably would break things so deferring to you

File: Sources/Sentry/SentryClient.m [hunk 1]
```diff
@@ -168,11 +169,17 @@ - (SentryId *)captureError:(NSError *)error
 - (SentryEvent *)buildErrorEvent:(NSError *)error
 {
     SentryEvent *event = [[SentryEvent alloc] initWithLevel:kSentryLevelError];
-    NSString *errorCodeAsString = [NSString stringWithFormat:@"%ld", (long)error.code];
-    SentryException *sentryException = [[SentryException alloc] initWithValue:errorCodeAsString
+    NSString *exceptionValue = [NSString stringWithFormat:@"Code: %ld LocalizedDescription: %@",
```
Comment (bruno-garcia) [inline Sources/Sentry/SentryClient.m:172]: localizedDescription will vary by locale, is this what we want to do?

File: Sources/Sentry/SentryClient.m [hunk 2]
```diff
@@ -168,11 +169,17 @@ - (SentryId *)captureError:(NSError *)error
...
+                                         (long)error.code, error.localizedDescription];
+    SentryException *exception = [[SentryException alloc] initWithValue:exceptionValue
                                                                          type:error.domain];
-    event.exceptions = @[ sentryException ];
-    event.fingerprint = @[ error.domain, [NSString stringWithFormat:@"%ld", (long)error.code] ];
+
+    // Sentry uses the error on the mechanism for gouping
+    SentryMechanism *mechanism = [[SentryMechanism alloc] initWithType:@"NSError"];
+    mechanism.error = error;
```
Comment (bruno-garcia) [inline Sources/Sentry/SentryClient.m:179]: Oh I was wonder what this was :)

https://develop.sentry.dev/sdk/event-payloads/exception/#ns_error

Comment (philipphofmann) [reply Sources/Sentry/SentryClient.m:179]: This is brand new 🔥😁

File: Sources/Sentry/SentryClient.m (hunk 1 above)
Comment (philipphofmann) [reply Sources/Sentry/SentryClient.m:172]: @mitsuhiko advocated for this as we also do this on other SDKs. I changed the logic a bit so we only add it if it exists in the userInfo.

File: Sources/Sentry/Public/SentryMechanism.h [hunk 3]
```diff
@@ -27,6 +27,11 @@ SENTRY_NO_INIT
  */
 @property (nonatomic, strong) NSDictionary<NSString *, id> *_Nullable data;
 
+/**
+ * Sentry uses the NSErrors domain and code for grouping. Only domain and code are serialized.
+ */
+@property (nonatomic, copy) NSError *_Nullable error;
```
Comment (bruno-garcia) [inline Sources/Sentry/Public/SentryMechanism.h:33]: Lets have a Sentry type here instead

File: Sources/Sentry/SentryClient.m [hunk 4]
```diff
@@ -168,13 +169,37 @@ - (SentryId *)captureError:(NSError *)error
...
-    [self setUserInfo:error.userInfo withEvent:event];
+    NSMutableString *exceptionValue =
+        [NSMutableString stringWithFormat:@"Code: %ld", (long)error.code];
+
+    // We only want to add the NSLocalizedDescription if the error contains one. As
+    // NSError.localizedDescription returns a default string contructed from the domain and code if
+    // the user info dictionary doesn’t contain a value for NSLocalizedDescriptionKey, we don't use
+    // NSError.localizedDescription as it would just again append domain and code, which we already
+    // have in the SentryException.
+    NSString *localizedDescription = error.userInfo[NSLocalizedDescriptionKey];
```
Comment (bruno-garcia) [inline Sources/Sentry/SentryClient.m:181]: This is what we used to do and after customer feedback we decided to drop it.
Probably better to put this in `exception.data` or `exception.mechanism.data` only.

**Review by bruno-garcia — COMMENTED** [review]
As discussed on our call

**Review by bruno-garcia — COMMENTED** [review]
LGTM!

**Pull request merged**